analytic_engine/src/sst/file_reader.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    ops::Range,
+    sync::{Arc, RwLock},
+};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use common_util::error::{GenericError, GenericResult};
+use object_store::{ObjectStoreRef, Path};
+
+pub type FileChunkReaderRef = Arc<dyn FileChunkReader>;
+
+#[async_trait]
+pub trait FileChunkReader: Send + Sync {
+    async fn file_size(&self) -> GenericResult<usize>;
+
+    async fn get_byte_range(&self, range: Range<usize>) -> GenericResult<Bytes>;
+
+    async fn get_byte_ranges(&self, ranges: &[Range<usize>]) -> GenericResult<Vec<Bytes>>;
+}
+
+/// A [`FileChunkReader`] implementation based on [`ObjectStore`].
+pub struct FileChunkReaderOnObjectStore {
+    path: Path,
+    store: ObjectStoreRef,
+    cached_file_size: RwLock<Option<usize>>,
+}
+
+impl FileChunkReaderOnObjectStore {
+    pub fn new(path: Path, store: ObjectStoreRef) -> Self {
+        Self {
+            path,
+            store,
+            cached_file_size: RwLock::new(None),
+        }
+    }
+}
+
+#[async_trait]
+impl FileChunkReader for FileChunkReaderOnObjectStore {
+    async fn file_size(&self) -> GenericResult<usize> {
+        // check cached filed_size first
+        {
+            let file_size = self.cached_file_size.read().unwrap();
+            if let Some(s) = file_size.as_ref() {
+                return Ok(*s);
+            }
+        }
+
+        // fetch the size from the underlying store
```

Comment:
double check here?

Reply:
A double check requires head fetch from oss with the lock held, which may lead to poor performance.